extension/js/background_page/background_page.ts

```diff
@@ -41,6 +41,7 @@ opgp.initWorker({ path: '/lib/openpgp.worker.js' });
 
   try {
     db = await ContactStore.dbOpen(); // takes 4-10 ms first time
+    await moveContactsToEmailsAndPubkeys(db);
```

Comment:
@tomholub this is the piece of code I was referring to when talking about contacts migration.
It would be more convenient to perform migration inside `dbOpen()` call. Thus we'll be able to delete `contacts` store right after successful conversion to `emails` and `pubkeys`.
We can, for example, pass a list of functions to `dbOpen` to nicely inject transformations so they don't clutter `ContactStore` code?

Reply:
or, perhaps, `ContactStore.update` will cause errors when called from inside upgrade transaction

Reply:
We can do all of this in `ContactStore` itself, no need to bring the migration out to here.

Reply:
> We can do all of this in ContactStore itself, no need to bring the migration out to here.

Looks like there is a problem with this approach. IDTransaction docs say that transaction is automatically completed when it goes out of scope, and this disallows usage of async function in `onupgradeneeded` etc.. They are fired alright, but the transaction is closed by the time the actual processing occurs (before waiting for the promise returned by `onupgradeneeded` handler to be resolved). I searched the github for projects where an async function is used as event handler, found only a small project https://github.com/SourceCodeBot/crudodb (I suspect they haven't tested it properly).
The general approach for transactional operations is (though I haven't found these exact requirement in IndexedDB specs) -- a handler must immediately open a new request, otherwise the transaction will be auto-completed.
We can't convert contacts to pubkeys/emails in a completely synchronized way because of async `KeyUtil.parse` function.
(This is the point for another discussion -- KeyUtil.parse is async only because it uses a Stream API approach for dearmoring of potentially large messages. As the keys are small, we may think about a sync implementation of KeyUtil.parse)
Still, maybe it's user-unfriendly to "hang" (will it hang?) the browser window for the process of database upgrade, converting multiple contacts, so it makes sense to keep the conversion out of the upgrade handler. The drawback is that the empty `contacts` table will be lying around until the next upgrade.

Reply:
>  Still, maybe it's user-unfriendly to "hang" (will it hang?) the browser window for the process of database upgrade, converting multiple contacts, so it makes sense to keep the conversion out of the upgrade handler. The drawback is that the empty contacts table will be lying around until the next upgrade.

It will possibly hang the whole browser when we do something like this in background page. Also depends on contacts count. For 50 contacts, not a problem. For 1000 contacts - possible problem.

I think our most reasonable chance is to do a migration of database structure itself, fire off async migration of the actual database, and hope for the best.

If the actual migration does not finish successfully, then on next migration it will be retried (like a browser restart). I think that's ok as long as there is this retry in place.

Best would be to then not migrate all contacts in one gigantic transaction, but one by one. If one of the transactions throws an error because whatever (parsing, concurency, etc), it would not fail the whole migration. 100 keys already migrated, key 101 failed, 50 more to go - the failed one gets catched and skipped, and it will chug along for the remaining 50 keys. Then next time it only has to retry one.

Reply:
I have implemented migration in batches of 50, but it's going to throw on an error and crash the extension.
If we skip, this portion of contacts won't be available for search -- is this what you're after?

Reply:
When there is an error, it's better to skip a batch then stop half way through the migration.

Reply:
Will the skipped batch get retried (without any special retry logic) next time I restart the browser?

Reply:
The conversion stop at the failed batch with the current implementation. 
as `contacts` store objects are deleted within the same transaction as populating new stores, the contacts from the failed batch will remain in the database, thus they will be retried on subsequent browser restart.